sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -0,0 +1,1191 @@
+`free` node is executed and after the `malloc` node is executed.
+
+Table 7. Member functions of the `command_graph` class (memory operations).
+[cols="2a,a"]
+|===
+|Member function|Description
+
+|
+[source,c++]
+----
+using namespace ext::oneapi::experimental;
+node add_malloc_device(void*& data, size_t numBytes, const property_list& propList = {});
+----
+|Adding a node that encapsulates a memory allocation operation.
+
+Preconditions:
+
+* This member function is only available when the `command_graph` state is
+  `graph_state::modifiable`.
+
+Parameters:
+
+* `data` - Return parameter set to the address of memory allocated.
+
+* `numBytes` - Size in bytes to allocate.
+
+* `propList` - Zero or more properties can be provided to the constructed node
+  via an instance of `property_list`.
+
+Returns: The memory allocation node which has been added to the graph
+
+Exceptions:
+
+* Throws synchronously with error code `invalid` if a queue is recording
+  commands to the graph.
+
+|
+[source,c++]
+----
+using namespace ext::oneapi::experimental;
+node add_free(void* data, const property_list& propList = {});
+----
+|Adding a node that encapsulates a memory free operation.
+
+Preconditions:
+
+* This member function is only available when the `command_graph` state is
+  `graph_state::modifiable`.
+
+Parameters:
+
+* `data` - Address of memory to free.
+
+* `propList` - Zero or more properties can be provided to the constructed node
+  via an instance of `property_list`.
+
+Returns: The memory freeing node which has been added to the graph.
+
+Exceptions:
+
+* Throws synchronously with error code `invalid` if a queue is recording
+  commands to the graph.
+
+|===
+
+Table 8. Member functions of the `command_graph` class (executable graph update).
+[cols="2a,a"]
+|===
+|Member function|Description
+
+|
+[source, c++]
+----
+using namespace ext::oneapi::experimental;
+void command_graph<graph_state::executable> update(const command_graph<graph_state::modifiable>& graph);
+----
+
+|Updates the executable graph node inputs & outputs from a topologically
+identical modifiable graph. The effects of the update will be visible
+on the next submission of the executable graph without the need for additional
+user synchronization. No changes to commands themselves will occur, such as to
+updating kernel or host task code to match that of the modifiable graph.
+
+Preconditions:
+
+* This member function is only available when the `command_graph` state is
+  `graph_state::executable`.
+
+Parameters:
+
+* `graph` - Modifiable graph object to update graph node inputs & outputs with.
+  This graph must have the same topology as the original graph used on
+  executable graph creation.
+
+Exceptions:
+
+* Throws synchronously with error code `invalid` if the topology of `graph` is
+  not the same as the existing graph topology, or if the nodes were not added in
+  the same order.
+|===
+
+=== Queue Class Modifications
```

Comment:
I have concerns about the "record and replay" API, mostly because the queue does not behave like a normal queue when it is in "record" mode.  Listing the differences:

* The submit functions return a "fake" event which is immediately in the "complete" state.
* The buffer destructor no longer blocks in the same way when the queue contains commands that uses accessors to the buffer.
* Of course, the commands are not executed.

By adding a new "record state" to the queue, I think you will promote a usage where the user changes the queue's initialization code to enter the "record state" and then passes the queue to some existing code.  Of course, this won't work because of the differences I list above.  What's worse, is that the code doesn't fail in any obvious way.  For example, attempting to wait on an event silently works even though the event is "fake".

I think we should expose "record and replay" in a different way.  Here are two options that come to mind:

* We could create some new class that is different from `queue`, but with similar member functions.  For example, this new class would still have a `submit` function, but it would not return an `event`.  This will trigger a compilation error if the user attempts to port existing code to this new "record queue", and the compilation error will prompt them to change the code such that it no longer uses the event.  In general, this approach would force users to look at all the code which adds commands to the queue, and make sure it will work in record mode.

* We could add a "record" mode to the standard SYCL `queue`, but the semantic would be that commands added to the queue would still be submitted to the device, and they would execute as normal.  Thus, the application could execute commands on a device while also recording them.  This would allow you to preserve all the existing semantics of the standard SYCL queue even while recording because the recorded commands would also be executed.